flycheck.el

```diff
@@ -7647,9 +7647,35 @@ Return the absolute path to the directory"
   "Return directory where rebar.config is located."
   (locate-rebar3-project-root buffer-file-name))
 
+(flycheck-def-option-var flycheck-erlang-rebar3-profile nil erlang-rebar3
+  "The rebar3 profile to use.
+
+The profile used when compiling, if VALUE is nil `test' will be used
+when the file is located in test directory, otherwise `default' will be
+used as profile."
+  :type '(choice
+          (const :tag "Guess rebar3 profile" nil)
+          (const :tag "Use default rebar3 profile" default)
+          (const :tag "Use test rebar3 profile" test))
+  :safe (lambda (value) (or (not value) (memq value '(default test))))
+  :package-version '(flycheck . "0.24"))
+
+(defun rebar3-profile ()
+  "Return rebar3 profile.
+
+Use flycheck-erlang-rebar3-profile if set, otherwise use test profile if
+dirname is test or else default."
+  (cond
```

Comment:
I don't think this does what you want.  `cond` is a like a `switch`.  Here you are saying: "Test the value of `(flycheck-erlang-rebar3-profile  (symbol-name flycheck-erlang-rebar3-profile))` (which will error, since `flycheck-erlang-rebar3-profile` is not a function), then if `(string= "test" ...` is true (which is unrelated to the value of the option), use `"test"`, if not, use `"default"`".

Did you test that code?

You could use `cond`, but there is an even simpler idiom.  Here is what you want:
```elisp
(or flycheck-erlang-rebar3-profile
    (and (string= "test" (file-name-base (directory-file-name (file-name-directory buffer-file-name))))
         "test")
    "default")
```
`or` and `and` are short-cicuiting and return the first (resp. last) non-nil value.  If `flycheck-erlang-rebar3-profile` is non-nil, then `or` returns it.  If it is nil, then we try to auto-detect.  If the `string=` succeeds (non-nil), then we return `"test"`.  If not, then `and` fails and we fallback to `"default"` (which is non-nil, and returned by `or`).

Also, `buffer-file-name` can be nil (some buffers have no file), so I would add:
```elisp
(and buffer-file-name
     (string= "test" (file-name-base (directory-file-name (file-name-directory buffer-file-name))))
     "test")
```

Please double-check that code by testing it on your setup.

Reply:
I did test the code but everything work with "test" profile so that might fail to show errors in my code :-)

I expected the cond clause to be (CONDITION BODY) i.e. flycheck-erlang-rebar3-profile to be interpreted as the CONDITION there, i.e. not nil will evaluate the BODY.

Also I want strings to be returned from the function and flycheck-erlang-rebar3-profile is set to a  symbol? (atom in my world) will that work with your 'or' alternative is there some auto-conversion in lisp?

Reply:
Hmm I tested some more and from what I can see it works as I expect it to do.

Reply:
> I expected the cond clause to be (CONDITION BODY) i.e. flycheck-erlang-rebar3-profile to be interpreted as the CONDITION there, i.e. not nil will evaluate the BODY.

You are right.   I don't know what I was thinking there.  The `cond` does what you want, and my suggestion doesn't (it's missing `symbol-name`).  I apologize for the brain fart.

> will that work with your 'or' alternative is there some auto-conversion in lisp?

There's no implicit conversion, so no it wouldn't work as is.  On the other hand we could also use strings instead of symbols for the option, but symbols are fine.

Reply:
Ah wait, do we want the user to be able to provide an arbitrary profile beyond `default` and `test` or not?

Reply:
phwee I was mightily embarrassed there for not testing enough, I hate when I get untested PR's :-)
Anyway thanks for your patience, I write lisp every 5 years or so, so stumble around.

Reply:
Ahh sorry for the embarrassment :(